### PR TITLE
feat(api): Add Movie Database API integration

### PR DESCRIPTION
**Pull request by maciejopalinski**

### CODE REVIEW

**Review by Mariusz-web — APPROVED** [review]

File: src/services/CacheManager.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        let url = this.getCacheFilename(api, params);
+
+        console.debug(`Trying to fetch data from cache... ${url}`);
+
+        if (method != Http.GET) {
+            throw new Error('HTTP method other than GET is not supported.');
+        } else {
+            try {
+                let resp = await fetch(url);
+                let json = await resp.json();
```
Comment (Veryyapeee) [inline src/services/CacheManager.ts]: const

File: src/services/CacheManager.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    async getCache(method: Http, api: ApiType, params: string) {
+        let url = this.getCacheFilename(api, params);
+
+        console.debug(`Trying to fetch data from cache... ${url}`);
+
+        if (method != Http.GET) {
+            throw new Error('HTTP method other than GET is not supported.');
+        } else {
+            try {
+                let resp = await fetch(url);
```
Comment (Veryyapeee) [inline src/services/CacheManager.ts]: const

File: src/services/CacheManager.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+        this.downloadFile(basename, JSON.stringify(json));
+    }
+
+    async getCache(method: Http, api: ApiType, params: string) {
+        let url = this.getCacheFilename(api, params);
+
+        console.debug(`Trying to fetch data from cache... ${url}`);
+
+        if (method != Http.GET) {
```
Comment (Veryyapeee) [inline src/services/CacheManager.ts]: `!==`

File: src/services/CacheManager.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        console.warn(`New cache entry!`, {
+            filename,
+            content: JSON.stringify(json)
+        });
+
+        this.downloadFile(basename, JSON.stringify(json));
+    }
+
+    async getCache(method: Http, api: ApiType, params: string) {
+        let url = this.getCacheFilename(api, params);
```
Comment (Veryyapeee) [inline src/services/CacheManager.ts]: const

File: src/services/CacheManager.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        element.style.display = 'none';
+        document.body.appendChild(element);
+
+        element.click();
+
+        document.body.removeChild(element);
+    }
+
+    reportNewCacheEntry(json: object, api: ApiType, params: string) {
+        let filename = this.getCacheFilename(api, params);
```
Comment (Veryyapeee) [inline src/services/CacheManager.ts]: const

File: src/services/CacheManager.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+import { ApiType, Http } from '../interfaces';
+
+class CacheManager {
+
+    getCacheFilename(api: ApiType, params: string) {
+        return ApiService.getURL(`/static/api/${api}/${params}.json`, '');
+    }
+
+    downloadFile(filename: string, text: string) {
+        let element = document.createElement('a');
```
Comment (Veryyapeee) [inline src/services/CacheManager.ts]: const

File: src/services/ApiService.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+            return json;
+        } catch (e) {
+            console.error(e);
+            throw new Error('Error occurred while fetching the API!');
+        }
+    }
+
+    async request(method: Http, api: ApiType, params: string, optional_params?: string) {
+        let cache_entry = await CacheManager.getCache(method, api, params);
```
Comment (Veryyapeee) [inline src/services/ApiService.ts]: const

File: src/services/ApiService.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+        try {
+            let resp = await fetch(url, {
+                method: method,
+                headers: {
+                    'X-RapidAPI-Host': API_BASE_URL.replace('https://', ''),
+                    'X-RapidAPI-Key': API_SECRET_KEY
+                }
+            });
+            let json = await resp.json();
```
Comment (Veryyapeee) [inline src/services/ApiService.ts]: const

File: src/services/ApiService.ts [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        return url;
+    }
+
+    async callAPI(method: Http, api: ApiType, params: string, optional_params?: string) {
+        let url = this.getURL(`/${api}/${process.env.API_SECRET_KEY}/${params}/${optional_params}`);
+
+        console.debug(`Trying to fetch data from API... ${url}`);
+
+        try {
+            let resp = await fetch(url, {
```
Comment (Veryyapeee) [inline src/services/ApiService.ts]: const and also name should be "response", meaningfull variables :P

File: src/services/ApiService.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    getURL(endpoint: string, prefix = API_BASE_URL) {
+        let url = prefix + endpoint;
+        while (url.match('//')) {
+            url = url.replace('//', '/');
+        }
+        return url;
+    }
+
+    async callAPI(method: Http, api: ApiType, params: string, optional_params?: string) {
+        let url = this.getURL(`/${api}/${process.env.API_SECRET_KEY}/${params}/${optional_params}`);
```
Comment (Veryyapeee) [inline src/services/ApiService.ts]: const

File: src/providers/ApiProvider.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
+import ApiService from '../services/ApiService';
+import { ApiType, Http, SearchData, TitleData } from '../interfaces';
+
+class ApiProvider {
+    search(q: string): Promise<SearchData> {
```
Comment (Veryyapeee) [inline src/providers/ApiProvider.ts]: in this case I don't know what does the "q" means, variable names should be meaningfull

File: src/interfaces/api/index.ts [hunk 12]
```diff
@@ -0,0 +1,43 @@
+export * from './Search';
+export * from './Title';
+
+export enum ApiType {
```
Comment (Veryyapeee) [inline src/interfaces/api/index.ts]: same here, I think that we won't need all of these endpoints

File: src/interfaces/api/Title.ts [hunk 13]
```diff
@@ -0,0 +1,161 @@
+// Generated by https://quicktype.io
+
+export interface TitleData {
```
Comment (Veryyapeee) [inline src/interfaces/api/Title.ts]: It's generated, but we don't need all of the properties, stick to the ones that we need

File: src/index.js [hunk 14]
```diff
@@ -1,7 +1,6 @@
 import 'regenerator-runtime/runtime'; //async/await with Parcel
 import { App } from './app/App';
 
-const BASE_API_URL = process.env.API_BASE_URL || 'https://google.com';
-
-window.onload = () =>
-  App({ renderOn: '#movie-app', options: { apiUrl: BASE_API_URL } });
+window.onload = () => {
```
Comment (Veryyapeee) [inline src/index.js:4]: unnecessary change, the => App({}) is shorter

File: src/app/App.js [hunk 15]
```diff
@@ -1,9 +1,24 @@
...
+            ${movies.reduce((r, movie) => r + `<li>${movie}</li>`, '')}
+        </ol>
+    </div>
 `;
 
-export const App = ({ renderOn, options }) => {
-  render({ on: renderOn, html: template });
+export const App = async ({ renderOn }) => {
+    let movies = [];
+    let data = await ApiProvider.search('Fast & Furious');
```
Comment (Veryyapeee) [inline src/app/App.js]: const instead of let, also
`const {results} = await ApiProvider.search(...)`

File: src/app/App.js [hunk 16]
```diff
@@ -1,9 +1,24 @@
...
+        <ol>
+            ${movies.reduce((r, movie) => r + `<li>${movie}</li>`, '')}
+        </ol>
+    </div>
 `;
 
-export const App = ({ renderOn, options }) => {
-  render({ on: renderOn, html: template });
+export const App = async ({ renderOn }) => {
+    let movies = [];
```
Comment (Veryyapeee) [inline src/app/App.js]: const instead of let

File: README.md [hunk 17]
```diff
@@ -1,7 +1,16 @@
...
 
-This project was made due to the **[CodersCamp2021](https://www.coderscamp.edu.pl/)**  
+This project was made due to the **[CodersCamp2021](https://www.coderscamp.edu.pl/)**
 
-You can read full documentation under this **[link](https://gracious-neumann-544c01.netlify.app/First%20project/participants)**  
+You can read full documentation under this **[link](https://gracious-neumann-544c01.netlify.app/First%20project/participants)**
 
-Application demo is available **[here](https://quizzical-feynman-5a4016.netlify.app/)**
\ No newline at end of file
+Application demo is available **[here](https://quizzical-feynman-5a4016.netlify.app/)**
```
Comment (Veryyapeee) [inline README.md:7]: This should be included in the documentation project

**Review by Veryyapeee — CHANGES_REQUESTED** [review]
After changing something let the reviewer know in the comment, so they can resolve conversation

**Comment by Veryyapeee** [comment]
Generally we could use the json-server for storing the data from API in case of limited calls instead of this cache logic, but I think we can put that ticket on the board for future refactor if we will have time

File: src/index.js (hunk 14 above)
Comment (maciejopalinski) [reply src/index.js:4]: Prettier did this

File: src/interfaces/api/index.ts (hunk 12 above)
Comment (maciejopalinski) [reply src/interfaces/api/index.ts]: Yeah maybe, but I think it would be better to have them just in case. Frontend won't be able to access those methods directly. They have only access to ApiProvider class that only has methods that we use.

**Comment by maciejopalinski** [comment]
@Veryyapeee I resolved almost all changes that you requested. I don't think we should remove unnecessary properties from response interfaces.

File: src/app/App.js (hunk 16 above)
Comment (Veryyapeee) [reply src/app/App.js]: I don't see const here :P

File: src/app/App.js [hunk 18]
```diff
@@ -1,9 +1,24 @@
...
+    const { results } = await ApiProvider.search('Fast & Furious');
+    console.debug(data);
```
Comment (Veryyapeee) [inline src/app/App.js]: Do we need this now?

File: src/index.js (hunk 14 above)
Comment (Veryyapeee) [reply src/index.js:4]: Pobably it's caused by your local config of prettier or eslint, prettier config in project doesn't have such a rule and we don't have an eslint configured :P  Minor

File: src/app/App.js (hunk 18 above)
Comment (maciejopalinski) [reply src/app/App.js]: It is just an example, but ok imma remove it

**Review by Veryyapeee — APPROVED** [review]
